Source/Error.swift

```diff
@@ -24,41 +24,428 @@
 
 import Foundation
 
-/// The domain used for creating all Alamofire errors.
-public let ErrorDomain = "org.alamofire.error"
-
-/// The custom error codes generated by Alamofire.
-public enum ErrorCode: Int {
-    case inputStreamReadFailed           = -6000
-    case outputStreamWriteFailed         = -6001
-    case contentTypeValidationFailed     = -6002
-    case statusCodeValidationFailed      = -6003
-    case dataSerializationFailed         = -6004
-    case stringSerializationFailed       = -6005
-    case jsonSerializationFailed         = -6006
-    case propertyListSerializationFailed = -6007
+/// `AFError` is the error type returned by the Alamofire framework. It encompasses several
+/// different types of errors, each with their own associated reasons.
+///
+/// - multipartEncodingFailed:          Returned by the multipart body data APIs. Contains a `MultipartEncodingFailureReason`.
+/// - inputStreamReadFailed:            <#inputStreamReadFailed description#>
+/// - outputStreamWriteFailed:          <#outputStreamWriteFailed description#>
+/// - responseValidationFailed:         <#responseValidationFailed description#>
+/// - responseSerializationFailed:      <#responseSerializationFailed description#>
+public enum AFError: Error {
+    
+    public enum MultipartEncodingFailureReason {
+        case notAFile(at: URL)
+        case failedToExtractFilename(for: URL)
+        case fileNotReachable(at: URL)
+        case isDirectory(at: URL)
+        case failedToFetchAttributes(from: URL)
+        case failedToCreateInputStream(from: URL)
+    }
+    
+    /// `OutputStreamFailureReason` is the reason returned for an `AFError.multipartEncodingFailed` error.
+    ///
+    /// - writeFailed:          <#writeFailed description#>
+    /// - failedToCreateStream: <#failedToCreateStream description#>
+    /// - fileAlreadyExists:    <#fileAlreadyExists description#>
+    /// - invalidFile:          <#invalidFile description#>
+    public enum OutputStreamFailureReason {
+        case writeFailed(to: OutputStream)
+        case failedToCreateStream(at: URL)
+        case fileAlreadyExists(at: URL)
+        case invalidFile(at: URL)
+    }
+    
+    public enum SerializationFailureReason {
+        case inputDataNil
+        case inputDataNilOrZeroLength
+        case invalidStringEncoding(String.Encoding)
+    }
+    
+    public enum ValidationFailureReason {
+        case contentTypeMismatch(acceptable: [String], actual: String)
+        case contentTypeMissingAndNotWildcard
+        case unacceptableStatusCode(code: Int)
+    }
+    
+    case multipartEncodingFailed(reason: MultipartEncodingFailureReason)
+    case inputStreamReadFailed(inputStream: InputStream)
+    case outputStreamWriteFailed(reason: OutputStreamFailureReason)
+    case responseValidationFailed(reason: ValidationFailureReason)
+    case responseSerializationFailed(reason: SerializationFailureReason)
+}
```

Comment:
I think this is an AWESOME start @jshier. I spent some time working through an alternate version of these with some different groupings.

```swift
public enum AFError: Error {
    public enum MultipartEncodingFailureReason {
        case bodyPartNotFile(at: URL)
        case bodyPartFilenameInvalid(at: URL)
        case bodyPartFileNotReachable(at: URL)
        case bodyPartFileIsDirectory(at: URL) // should probably lump together with `.bodyPartNotFile`
        case bodyPartAttributesNotAvailable(at: URL)
        case bodyPartInputStreamCreationFailed(at: URL)

        case outputStreamCreationFailed(at: URL)
        case outputStreamFileAlreadyExists(at: URL)
        case outputStreamFileIsInvalid(at: URL)
        case outputStreamWriteFailed(for: OutputStream)
    }

    public enum ValidationFailureReason {
        case missingContentType(accepted: [String])
        case unacceptableContentType(accepted: [String], response: String)
        case unacceptableStatusCode(code: Int)
    }

    public enum SerializationFailureReason {
        case inputDataNil
        case inputDataZeroLength
        case stringSerializationFailed(encoding: String.Encoding)
        case jsonSerializationFailed(error: NSError)
        case propertyListSerializationFailed(error: NSError)
    }

    case requestFailed(error: NSError) // Not sure if we want to wrap all NSURLSession errors (seems like a good idea)
    case multipartEncodingFailed(reason: MultipartEncodingFailureReason)
    case validationFailed(reason: ValidationFailureReason)
    case serializationFailed(reason: SerializationFailureReason)
}
```

Would love to see what you think here. I made the `MultipartEncoding` errors MUCH more specific and lumped the `.multipartEncodingFailed`, `.inputStreamReadFailed` and `.outputStreamWriteFailed` errors all into one. They don't really stand on their own IMO.

Then I reworked the other two a bit. I think we should definitely wrap JSON and plist serialization errors. I even think we should seriously consider wrapping all request / network errors inside an `AFError` as well. That would make it MUCH easier for the consumer to switch on the overall error type. Most consumers won't probably need to dig into the reason, but could leverage the top four buckets.